src/core/dom/jsx.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * @fileoverview
+ * Minimal implementation of JSX that outputs DOM nodes.
+ *
+ * Usage:
+ *   import * as Preact from '#core/dom/jsx';
+ *
+ * This library is nicer than templates in bundles that do not include Preact,
+ * but it does not attempt to implement JSX fully.
+ *
+ * These features are omitted for the sake of bundle size:
+ *
+ * 🚫 Attribute names are not re-mapped.
+ *   - Use standard HTML attribute names on elements (`class`, not `className`).
+ *   - Dashes are ok, like `data-foo="bar"`.
+ *
+ * 🚫 No objects in attribute values, like `style={{width: 40}}`
+ *   - For `style`, use strings instead.
+ *   - For `class`, call `objstr()` or use strings instead.
+ *
+ * 🚫 No Fragments
+ *   - Instead use a root node, or split into an array of nodes.
+ *
+ * 🚫 No dangerouslySetInnerHTML
+ *   - You should not do this anyway.
+ */
+import {devAssert} from '#core/assert';
+
+/**
+ * @param {!Element} parent
+ * @param {*} child
+ */
+function appendChild(parent, child) {
+  if (child === false || child == null) {
+    return;
+  }
+  if (Array.isArray(child)) {
+    child.forEach((child) => {
+      appendChild(parent, child);
+    });
+    return;
+  }
+  parent.appendChild(
+    child?.nodeType
+      ? /** @type {!Node} */ (child)
+      : self.document.createTextNode(String(child))
+  );
+}
+
+/**
+ * @param {!Element} element
+ * @param {string} name
+ * @param {*} value
+ */
+function setAttribute(element, name, value) {
+  if (value === false || value == null) {
+    return;
+  }
+  if (typeof value === 'function' && name[0] === 'o' && name[1] === 'n') {
+    const eventName = name.toLowerCase().substr(2);
+    element.addEventListener(eventName, value);
+    return;
+  }
+  element.setAttribute(name, value === true ? '' : String(value));
```

Comment:
Nit: We could probably get away with just `String(value)`, since presence of anything toggles the boolean attr state.

Reply:
Unfortunately this would get hairy with existing tests since they might check for `''` instead of `hasAttribute`.